### PR TITLE
Zig update: Update cast builtins to single-argument variants

### PR DESCRIPTION
**Pull request by ehaas**

### CODE REVIEW

File: src/Compilation.zig [hunk 1]
```diff
@@ -124,7 +124,7 @@ fn generateDateAndTime(w: anytype) !void {
 
     const day_names = [_][]const u8{ "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun" };
     // days since Thu Oct 1 1970
-    const day_name = day_names[@truncate(u8, (epoch_day.day + 3) % 7)];
+    const day_name = day_names[@as(u8, @truncate((epoch_day.day + 3) % 7))];
```
Comment (Vexu) [inline src/Compilation.zig:127]: Is this `@as` required?

File: src/Compilation.zig [hunk 2]
```diff
@@ -1297,10 +1297,10 @@ test "addSourceFromReader" {
             defer comp.deinit();
 
             var buf_reader = std.io.fixedBufferStream(str);
-            const source = try comp.addSourceFromReader(buf_reader.reader(), "path", @intCast(u32, str.len));
+            const source = try comp.addSourceFromReader(buf_reader.reader(), "path", @intCast(str.len));
 
             try std.testing.expectEqualStrings(expected, source.buf);
-            try std.testing.expectEqual(warning_count, @intCast(u32, comp.diag.list.items.len));
+            try std.testing.expectEqual(warning_count, @as(u32, @intCast(comp.diag.list.items.len)));
```
Comment (Vexu) [inline src/Compilation.zig:1303]: Shouldn't this get the type from `warning_count`?

File: src/Parser.zig [hunk 3]
```diff
@@ -3344,7 +3344,7 @@ fn findScalarInitializer(p: *Parser, il: **InitList, ty: *Type, actual_ty: Type,
         }
         const struct_il = il.*;
         while (index < fields.len) : (index += 1) {
-            const field = fields[@intCast(u32, index)];
+            const field = fields[@as(u32, @intCast(index))];
```
Comment (Vexu) [inline src/Parser.zig:3347]: Why was there a cast here in the first place?

File: src/Parser.zig [hunk 4]
```diff
@@ -3401,7 +3401,7 @@ fn findAggregateInitializer(p: *Parser, il: **InitList, ty: *Type, start_index:
 
         const field_count = struct_ty.data.record.fields.len;
         if (index < field_count) {
-            ty.* = struct_ty.data.record.fields[@intCast(usize, index)].ty;
+            ty.* = struct_ty.data.record.fields[@as(usize, @intCast(index))].ty;
```
Comment (Vexu) [inline src/Parser.zig:3404]: Ditto.

File: src/Source.zig [hunk 5]
```diff
@@ -45,15 +45,15 @@ pub fn lineCol(source: Source, loc: Location) LineCol {
     var start: usize = 0;
     // find the start of the line which is either a newline or a splice
     if (std.mem.lastIndexOfScalar(u8, source.buf[0..loc.byte_offset], '\n')) |some| start = some + 1;
-    const splice_index = for (source.splice_locs, 0..) |splice_offset, i| {
+    const splice_index: u32 = for (source.splice_locs, 0..) |splice_offset, i| {
         if (splice_offset > start) {
             if (splice_offset < loc.byte_offset) {
                 start = splice_offset;
-                break @intCast(u32, i) + 1;
+                break @as(u32, @intCast(i)) + 1;
```
Comment (Vexu) [inline src/Source.zig:52]: Is this `@as` needed?

File: test/fuzz/fuzz_lib.zig [hunk 6]
```diff
@@ -12,7 +12,7 @@ const Parser = aro.Parser;
 var fixed_buffer_mem: [10 * 1024 * 1024]u8 = undefined;
 
 export fn compile_c_buf(buf: [*]const u8, len: c_int) void {
-    compileSlice(buf[0..@intCast(usize, len)]) catch unreachable;
+    compileSlice(buf[0..@as(usize, @intCast(len))]) catch unreachable;
```
Comment (Vexu) [inline test/fuzz/fuzz_lib.zig:15]: Doesn't seem like `@as` should be needed.

File: src/Compilation.zig (hunk 2 above)
Comment (ehaas) [reply src/Compilation.zig:1303]: It looks like it's not able to see through the `@TypeOf(expected)` where `expected` is `anytype` - without the cast I get:

```
src/Compilation.zig:1303:56: error: @intCast must have a known result type
            try std.testing.expectEqual(warning_count, @intCast(comp.diag.list.items.len));
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/Compilation.zig:1303:56: note: result type is unknown due to anytype parameter
src/Compilation.zig:1303:56: note: use @as to provide explicit result type
```

Comment (Vexu) [reply src/Compilation.zig:1303]: That seems like a bug.

File: src/Compilation.zig (hunk 1 above)
Comment (ehaas) [reply src/Compilation.zig]: Removed the cast and the truncate

File: src/Parser.zig (hunk 4 above)
Comment (ehaas) [reply src/Parser.zig:3404]: `index` is `u64` so it's needed for 32-bit builds

File: src/Parser.zig (hunk 3 above)
Comment (ehaas) [reply src/Parser.zig:3347]: This one might not actually be needed; I'll push an update and see if it still builds

File: src/Source.zig (hunk 5 above)
Comment (ehaas) [reply src/Source.zig:52]: Compilation fails with unknown result type without it

File: src/Compilation.zig (hunk 1 above)
Comment (Vexu) [reply src/Compilation.zig]: Seems like the truncate is needed for 32-bit builds.

File: src/Parser.zig (hunk 4 above)
Comment (Vexu) [reply src/Parser.zig:3404]: Right, forgot about that.

Comment (Vexu) [reply src/Parser.zig:3404]: `@as` should still be unnecessary.

File: test/fuzz/fuzz_lib.zig (hunk 6 above)
Comment (ehaas) [reply test/fuzz/fuzz_lib.zig]: Removed. This file has bitrotted a bit; I haven't run the fuzzer in a while.

File: src/Parser.zig (hunk 4 above)
Comment (ehaas) [reply src/Parser.zig]: Removed this and the other one in `findScalarInitializer` and seems to have worked

File: src/Compilation.zig (hunk 1 above)
Comment (Vexu) [reply src/Compilation.zig:127]: The `@as` should still be unnecessary.

Comment (ehaas) [reply src/Compilation.zig:127]: Without it, on a 64bit system I get

```
src/Compilation.zig:127:32: error: destination type 'usize' has more bits than source type 'u47'
    const day_name = day_names[@truncate((epoch_day.day + 3) % 7)];
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/Compilation.zig:127:32: note: destination type has 64 bits
src/Compilation.zig:127:62: note: operand type has 47 bits
    const day_name = day_names[@truncate((epoch_day.day + 3) % 7)];
                                         ~~~~~~~~~~~~~~~~~~~~^~~
```

Comment (ehaas) [reply src/Compilation.zig:127]: Maybe we should just be using `@intCast` in the first place instead of truncate

**Pull request merged**